pkg/alertmanager/alertmanager.go

```diff
@@ -122,11 +136,22 @@ func New(cfg *Config, reg *prometheus.Registry) (*Alertmanager, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to create notification log: %v", err)
 	}
+
+	// We currently have 3 operational modes:
+	// 1) Alertmanager no clustering
+	// 2) Alertmanager clustering with upstream Gossip
+	// 3) Alertmanager sharding
+	// #1 and #2 are covered by the 1st conditional, then the 2nd conditional covers #3
 	if cfg.Peer != nil {
 		c := cfg.Peer.AddState("nfl:"+cfg.UserID, am.nflog, am.registry)
 		am.nflog.SetBroadcast(c.Broadcast)
 	}
 
+	if cfg.Peer == nil && cfg.ShardingEnabled {
+		c := am.state.AddState("nfl:"+cfg.UserID, am.nflog)
```

Comment:
Why do we need duplicated logic here? I'm wondering if we could refactor `cfg.Peer` to be an interface and we set it either to the alertmanager clustering peer or `am.state` so that we keep the same logic between the two, but the underlying implementation changes.

Reply:
I think that's very reasonable, however, we're not done just yet. Given the next step is to implement the:

- Full state replication
- Persistence of state to object storage

I feel like we should wait until those two lands so that we can have an idea of how the final interface looks like.

Reply:
Adding to this argument is the fact that `Peer`, as defined by upstream, is something that's global - it is always the same for _all_ the alertmanagers thus initialized in the multitenant manager. Whereas in this design `state` is something local to the tenent's alertmanager.

Reply:
I took this for a spin - is this better?